src/apps/SealedBidTokenSale.sol

```diff
@@ -0,0 +1,270 @@
+    bool public capReached;
+    /// @notice Running total of USDC deposited into sale
+    uint256 public totalDeposited;
+    /// @notice Merkle root for verifying token allocations
+    bytes32 public merkleRoot;
+    /// @notice Maps user addresses to their USDC deposit amounts
+    mapping(address => uint256) public deposits;
+    /// @notice Maps user addresses to whether they've claimed tokens
+    mapping(address => bool) public hasClaimed;
+
+    /* ============ Constructor ============ */
+
+    /**
+     * @notice Initializes the token sale with required parameters
+     * @param _saleToken Address of the token being sold
+     * @param _treasury Address where sale proceeds will be sent
+     * @param _usdcToken Address of the USDC token contract
+     * @param _startTime Timestamp when sale will begin
+     * @param _minimumCap Minimum USDC amount required for sale success
+     */
+    constructor(address _saleToken, address _treasury, address _usdcToken, uint256 _startTime, uint256 _minimumCap)
+        Ownable(msg.sender)
+    {
+        if (_saleToken == address(0)) revert InvalidSaleTokenAddress(_saleToken);
+        if (_treasury == address(0)) revert InvalidTreasuryAddress(_treasury);
+
+        saleToken = IERC20(_saleToken);
+        treasury = _treasury;
+        USDC = IERC20(_usdcToken);
+        startTime = _startTime;
+        minimumCap = _minimumCap;
+    }
+
+    /* ============ User Functions ============ */
+
+    /**
+     * @notice Allows users to deposit USDC into the token sale
+     * @dev - Sale must be active (after start time, before end)
+     *      - Amount must be greater than zero
+     *      - Updates user's deposit balance and total deposits
+     *      - Transfers USDC from user to contract
+     * @param amount Amount of USDC to deposit
+     */
+    function deposit(uint256 amount) external nonReentrant {
+        // Verify sale is active and deposit is valid
+        if (block.timestamp < startTime) revert SaleNotStarted(block.timestamp, startTime);
+        if (saleEnded) revert SaleAlreadyEnded(block.timestamp);
+        if (amount == 0) revert ZeroDeposit();
+
+        // Update deposit accounting
+        deposits[msg.sender] += amount;
+        totalDeposited += amount;
+
+        // Transfer USDC from user to contract
+        USDC.safeTransferFrom(msg.sender, address(this), amount);
+
+        // Emit deposit event
+        emit Deposited(msg.sender, amount);
+    }
+
+    /**
+     * @notice Allows users to withdraw their USDC if sale failed to reach minimum cap
+     * @dev - Sale must be ended and minimum cap not reached
+     *      - User must have deposited USDC
+     *      - Sends full deposit amount back to user
+     *      - Sets user's deposit balance to zero
+     */
+    function withdraw() external nonReentrant {
+        // Verify sale has ended unsuccessfully
+        if (!saleEnded) revert SaleNotEnded(block.timestamp);
+        if (capReached) revert CapReached();
```

Comment:
some users should be able to withdraw even if the cap was reached.

Their bids were not picked

Reply:
Unless you want to use the claimTokens for this, then its fine

Reply:
This will be done via `claimTokens` because they may need to withdraw partial amounts. Also, it creates a clear separation between two different outcomes (cap reached/not reached).